crates/adapters/src/server/mod.rs

```diff
@@ -530,13 +534,21 @@ async fn heap_profile() -> impl Responder {
 
 #[get("/dump_profile")]
 async fn dump_profile(state: WebData<ServerState>) -> impl Responder {
+    let (sender, receiver) = oneshot::channel();
     match &*state.controller.lock().unwrap() {
+        None => return Err(missing_controller_error(&state)),
         Some(controller) => {
-            controller.dump_profile();
-            Ok(HttpResponse::Ok().json("Profile dump initiated"))
+            controller.start_graph_profile(Box::new(move |profile| {
+                sender.send(profile).expect("failed to send profile");
```

Comment:
It's probably better to fail gracefully here, e.g., log an error. The situation I want to prevent is that it takes too long to generate the profile (e.g., because Feldera is stuck in `step()`), in the meantime the HTTP request timeouts and gets canceled. By the time the callback is invoked, the receiver is gone, so `sender.send()` fails. Not sure this is actually possible, but just to be safe.

Reply:
Good idea, done.